// singularci/dist/src/Targets/GitHubModule/tasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatePullDockerImageTask = exports.generateBuildDockerImageTask = exports.generateRunTask = exports.generateCheckoutTask = void 0;
const generateCheckoutTask = (task) => {
    return {
        uses: '@actions/checkout@v3',
        with: {
            repository: task.getRepositoryURL()
        }
    };
};
exports.generateCheckoutTask = generateCheckoutTask;
const generateRunTask = (task) => {
    let runCommand = "";
    if (task.getCommands().length == 1) {
        runCommand = task.getCommands()[0];
    }
    else if (task.getCommands().length > 1) {
        for (let i = 0; i < task.getCommands().length; i++) {
            runCommand += `${task.getCommands()[i]}\n`;
        }
    }
    return {
        run: runCommand
    };
};
exports.generateRunTask = generateRunTask;
const generateBuildDockerImageTask = (task) => {
    return {
        uses: "djbender/docker-buildx-pull-or-build@v0.5",
        with: {
            docker_username: task.getUserName(),
            docker_password: task.getPassword(),
            dockerfile: task.getBuildFilePath(),
            image: task.getImageName()
        }
    };
};
exports.generateBuildDockerImageTask = generateBuildDockerImageTask;
const generatePullDockerImageTask = (task) => {
    return {
        uses: "djbender/docker-buildx-pull-or-build@v0.5",
        with: {
            docker_username: task.getUserName(),
            docker_password: task.getPassword(),
            // will pull if dockerfile path is left empty
            dockerfile: "",
            image: task.getImageName()
        }
    };
};
exports.generatePullDockerImageTask = generatePullDockerImageTask;
